### PR TITLE
Classes in ScatteredArchive are processed only once.

### LINKED ISSUE #24415: [7.0.4] ServletContextListener.contextInitialized invoked multiple times
Hello,

In the context of a WAR, we have a simple `ServletContextListener` to initialize/shutdown our application when it is deployed. We only have **one single** servlet, so we expect the methods `contextInitialized` and `contextDestroyed` to be invoked only once when the application is deployed and undeployed.

Here is some piece of code:

```java
@WebListener
public class DispatcherListener implements ServletContextListener {
    private static final Logger LOG = Logger.getLogger(DispatcherListener.class.getName());

    @Override
    public void contextInitialized(ServletContextEvent sce) {
        log.log(Level.SEVERE, "Listener initialized");
    }
}
```

```java
@WebServlet(urlPatterns="/*", loadOnStartup=1)
public class DispatcherServlet extends HttpServlet {
   //nothing special here
}
```

Here the log output after the deployment:
```
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
SEVERE: Listener initialized
May 13, 2023 2:01:56 PM my.application.DispatcherListener contextInitialized
```

In Payara 6.2023.4, the log does only appear once, which is (in my sense) the expected behavior.

### Environment:

* JDK17
* glassfish-embedded-all 7.0.4
* Linux Debian

### PR DESCRIPTION
**Pull request by OndroMih**
They are not processed again when the parser is called for JARs (web fragments). Signed-off-by:Ondro Mihalyi 

Fixes #24415.

To do:

* write a test that covers #24415 with both ScatteredArchive and with a plain WAR file. The issue #24415 doesn't happen for a plain WAR file even without this fix. A reproducer needs to bundle at least 1 JAR file in the lib directory.
  * I will wait until https://github.com/eclipse-ee4j/glassfish/pull/24576 is merged and then add a test in the newly enabled GF embedded test suite
* Decide what to do for `WebFragmentDescriptor`. This solution returns no items for it. Maybe it should return all classes in the nested JAR file referenced from the descriptor.

### CODE REVIEW

**Comment by OndroMih** [comment]
Added test. Now looking at what to do with the WebFragmentDescriptor.

**Comment by OndroMih** [comment]
I won't have time to fix the WebFragmentDescriptor part any time soon. Since nobody reported any issue with it, I won't fix it in this PR. Instead, I've created a separate issue for it: https://github.com/eclipse-ee4j/glassfish/issues/24592.

This PR is now ready to be reviewed and merged.

**Comment by OndroMih** [comment]
@dmatej , @avpinchuk , @ctabin , please review.

File: appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java [hunk 1]
```diff
@@ -108,42 +135,91 @@ public void test() throws Exception {
...
+        logger.log(INFO, "***** SUCCESS **** Found [" + result + "] in the response.*****");
     }
 
-    private void get(String urlStr, String result) throws Exception {
-        URL url = new URL(urlStr);
+    private List<String> getLinesFromUrl(URL url) throws Exception {
         URLConnection yc = url.openConnection();
-        System.out.println("\nURLConnection [" + yc + "] : ");
+        logger.log(DEBUG, "\nURLConnection [" + yc + "] : ");
         BufferedReader in = new BufferedReader(new InputStreamReader(
```
Comment (avpinchuk) [inline appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:201]: return from inside try-with-resources?

File: appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java [hunk 2]
```diff
@@ -108,42 +135,91 @@ public void test() throws Exception {
...
+        List<String> inLines = getLinesFromUrl(new URL(urlStr));
+        String result = key + ":" + value;
+        MatcherAssert.assertThat("Output from servlet", inLines, hasItem(is(result)));
+        logger.log(INFO, "***** SUCCESS **** Found [" + result + "] in the response.*****");
     }
 
-    private void get(String urlStr, String result) throws Exception {
-        URL url = new URL(urlStr);
+    private List<String> getLinesFromUrl(URL url) throws Exception {
         URLConnection yc = url.openConnection();
```
Comment (avpinchuk) [inline appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:199]: Wrap with try and disconnect in finally?

Comment (OndroMih) [reply appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:199]: There's no disconnect on URLConnection. The connection isn't created with `URL.openConnection()`. It's created later, with `yc.getInputStream()`, see the Javadoc: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/URL.html#openConnection()

File: appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java [hunk 3]
```diff
@@ -108,42 +135,91 @@ public void test() throws Exception {
...
+        sa.addClassPath(new File(PROJECT_DIR, "target/test-dependencies/hamcrest.jar"));
+        sa.addClassPath(new File(PROJECT_DIR, "target/test-dependencies/junit-jupiter-engine.jar"));
 
+        URI warURI = sa.toURI();
+        printContents(warURI);
+
+        // Deploy archive
+        Deployer deployer = glassfish.getDeployer();
+        String appname = deployer.deploy(warURI);
+        System.out.println("Deployed [" + appname + "]");
```
[the remark this answers is not in the document]
Comment (OndroMih) [reply appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:169]: Fixed.

File: appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java (hunk 1 above)
Comment (OndroMih) [reply appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:201]: Fixed.

File: appserver/tests/embedded/scatteredarchive/src/main/webapp/WEB-INF/glassfish-web.xml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE glassfish-web-app PUBLIC "-//GlassFish.org//DTD GlassFish Application Server 3.1 Servlet 3.0//EN" "http://glassfish.org/dtds/glassfish-web-app_3_0-1.dtd">
 <!--
-
-    Copyright (c) 2010, 2018 Oracle and/or its affiliates. All rights reserved.
-
+    Copyright (c) 1997, 2018 Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (OndroMih) [reply appserver/tests/embedded/scatteredarchive/src/main/webapp/WEB-INF/glassfish-web.xml:4]: Fixed.

File: appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/contextInitialized/ContextInitializedTestServlet.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    public static final String LABEL_CONTEXT_INITIALIZED_COUNTER = "Number of times context was initialized";
+
+    @Override
+    protected void doGet(HttpServletRequest httpServletRequest,
+                         HttpServletResponse httpServletResponse) throws ServletException, IOException {
+        PrintWriter out = httpServletResponse.getWriter();
+        out.println(LABEL_CONTEXT_INITIALIZED_COUNTER + ":" + ApplicationStatus.contextInitializedCounter);
+        out.flush();
+        out.close();
```
[the remark this answers is not in the document]
Comment (OndroMih) [reply appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/contextInitialized/ContextInitializedTestServlet.java:39]: Fixed.

File: appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java (hunk 2 above)
Comment (avpinchuk) [reply appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:199]: You're right. Sorry :)

Comment (OndroMih) [reply appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java:199]: No problem. Thanks for being so thorough :)

**Review by avpinchuk — APPROVED** [review]

**Review by arjantijms — APPROVED** [review]

**Pull request merged**